File: framework/backend/kotlin/packages/k1s0-health/src/main/kotlin/dev/k1s0/health/HealthCheck.kt
package dev.k1s0.health

import kotlinx.serialization.Serializable

/**
 * Interface for health check probes used by Kubernetes liveness and readiness checks.
 */
public fun interface HealthCheck {
    /**
     * Performs the health check.
     *
     * @return The [HealthStatus] indicating the component's health.
     */
    public suspend fun check(): HealthStatus
}

/**
 * Health status of a single component or the overall service.
 *
 * @property name The name of the component being checked.
 * @property status Whether the component is healthy.
 * @property message Optional detail message.
 */
@Serializable
public data class HealthStatus(
    val name: String,
    val status: Status,
    val message: String? = null,
) {
    /** Health status values. */
    public enum class Status { UP, DOWN }
}

/** A liveness check that always returns UP (the process is running). */
public class LivenessCheck : HealthCheck {
    override suspend fun check(): HealthStatus =
        HealthStatus(name = "liveness", status = HealthStatus.Status.UP)
}

/**
 * A readiness check that aggregates multiple [HealthCheck] instances.
 *
 * Returns DOWN if any of the checks fail.
 */
public class ReadinessCheck(
    private val checks: List<HealthCheck>,
) : HealthCheck {
    override suspend fun check(): HealthStatus {
        val results = checks.map { it.check() }
        val allUp = results.all { it.status == HealthStatus.Status.UP }
        return HealthStatus(
            name = "readiness",
            status = if (allUp) HealthStatus.Status.UP else HealthStatus.Status.DOWN,
            message = if (allUp) null else results.filter { it.status == HealthStatus.Status.DOWN }
                .joinToString(", ") { "${it.name}: ${it.message ?: "DOWN"}" },
        )
    }
}

File: regions/system/library/swift/health/Sources/K1s0Health/HealthCheck.swift
import Foundation

public enum HealthStatus: String, Sendable {
    case healthy
    case degraded
    case unhealthy
}

public struct CheckResult: Sendable {
    public let status: HealthStatus
    public let message: String?

    public init(status: HealthStatus, message: String? = nil) {
        self.status = status
        self.message = message
    }
}

public struct HealthResponse: Sendable {
    public let status: HealthStatus
    public let checks: [String: CheckResult]
    public let timestamp: Date

    public init(status: HealthStatus, checks: [String: CheckResult], timestamp: Date) {
        self.status = status
        self.checks = checks
        self.timestamp = timestamp
    }
}

public protocol HealthCheck: Sendable {
    var name: String { get }
    func check() async throws
}
